test/http-agent.js

```diff
@@ -0,0 +1,268 @@
+const portastic = require('portastic');
+const proxy = require('proxy');
+const request = require('request');
+
+const { Server } = require('../src/index');
+const { TargetServer } = require('./utils/target_server');
+
+describe('HTTP Agent Support', () => {
+    let mainProxyServer;
+    let mainProxyServerPort;
+    let upstreamProxyServer;
+    let upstreamProxyPort;
+    let targetServer;
+    let targetServerUrl;
+
+    before(async () => {
+        // Get free ports
+        const freePorts = await portastic.find({ min: 50000, max: 50100 });
+
+        // Setup target server
+        const targetServerPort = freePorts.shift();
+        targetServer = new TargetServer({
+            port: targetServerPort,
+            useSsl: false,
+        });
+        await targetServer.listen();
+        targetServerUrl = `http://localhost:${targetServerPort}`;
+
+        // Setup upstream proxy server
+        upstreamProxyPort = freePorts.shift();
+        await new Promise((resolve, reject) => {
+            upstreamProxyServer = proxy(http.createServer());
+            upstreamProxyServer.listen(upstreamProxyPort, (err) => {
+                if (err) return reject(err);
+                resolve();
+            });
+        });
+
+        // Setup main proxy server with custom agents
+        mainProxyServerPort = freePorts.shift();
+    });
+
+    after(() => {
+        if (targetServer) targetServer.close();
+        if (upstreamProxyServer) upstreamProxyServer.close();
+        if (mainProxyServer) mainProxyServer.close(true);
+    });
+
+    it('accepts httpAgent and httpsAgent in prepareRequestFunction', async () => {
+        const httpAgent = new http.Agent({ keepAlive: true });
+        const httpsAgent = new https.Agent({ keepAlive: true });
+
+        let agentsPassed = false;
+
+        if (mainProxyServer) await mainProxyServer.close(true);
+
+        mainProxyServer = new Server({
+            port: mainProxyServerPort,
+            prepareRequestFunction: () => {
+                agentsPassed = true;
+                return {
+                    upstreamProxyUrl: `http://localhost:${upstreamProxyPort}`,
+                    httpAgent,
+                    httpsAgent,
+                };
+            },
+        });
+
+        await mainProxyServer.listen();
+
+        // Make HTTP request through the proxy
+        await new Promise((resolve, reject) => {
+            request({
+                url: `${targetServerUrl}/hello-world`,
+                proxy: `http://localhost:${mainProxyServerPort}`,
+            }, (error, response) => {
+                if (error) return reject(error);
+                expect(response.statusCode).to.eql(200);
+                resolve();
+            });
+        });
+
+        expect(agentsPassed).to.be.true;
+
+        // Cleanup agents
+        httpAgent.destroy();
+        httpsAgent.destroy();
+    });
+
+    it('reuses connections with keepAlive agents (sticky IP simulation)', async () => {
```

Comment:
What about instead of verifing implementation (hooks into 'free' event, inspect socket implementation details like remoteAddress and port, compare socketIds) to veirfy an observable effect of connection pooling (behavior)?

What we can do instead? Counts actual connections at the upstream proxy server. With this we can verify that fewer connections are made than requests (observable behavior of pooling).

What do you think?

Reply:
that's a great idea. Lemme dwell on it